include/nbl/builtin/hlsl/numeric_limits.hlsl

```diff
@@ -0,0 +1,202 @@
+// Copyright (C) 2022 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_NUMERIC_LIMITS_INCLUDED_
+#define _NBL_BUILTIN_HLSL_NUMERIC_LIMITS_INCLUDED_
+
+#include <type_traits.hlsl>
+
+using half = double;
+
+// C++ headers
+#ifndef __HLSL_VERSION
+#include <limits>
+#endif
+
+namespace nbl::hlsl::numeric_limits
+{
+
+#if 1 //def __HLSL_VERSION
+
+enum float_denorm_style {
+    denorm_indeterminate = -1,
+    denorm_absent        = 0,
+    denorm_present       = 1
+};
+
+enum float_round_style {
+    round_indeterminate       = -1,
+    round_toward_zero         = 0,
+    round_to_nearest          = 1,
+    round_toward_infinity     = 2,
+    round_toward_neg_infinity = 3
+};
+
+template<class T> struct num_base;
+
+template<class T> struct num_traits
+{
+    static const T MIN        = nbl::hlsl::type_traits::is_signed<T>::value ? ((T(1)<<num_base<T>::digits)-T(1)) : T(0);
+    static const T MAX        = (T(1)<<num_base<T>::digits)-T(1);;
+    static const T DENORM_MIN = 0;
+};
+
+template<> 
+struct num_traits<half>
+{
+    static const half MIN         = 6.103515e-05;
+    static const half MAX         = 65504;
+    static const half DENORM_MIN  = 5.96046448e-08;
+};
+
+template<> 
+struct num_traits<float>
+{
+    static const float MAX         = 3.402823466e+38F;
+    static const float MIN         = 1.175494351e-38F;
+    static const float DENORM_MIN  = 1.401298464e-45F;
+};
+
+template<> 
+struct num_traits<double>
+{
+    static const double MAX         = 1.7976931348623158e+308;
+    static const double MIN         = 2.2250738585072014e-308;
+    static const double DENORM_MIN  = 4.9406564584124654e-324;
+};
+
+
+template<class T>
+struct num_base
+{
```

Comment:
all this hackery is fine as long as you hide it in `impl` or `detail` extra namespace